ledger/src/shred/merkle.rs

```diff
@@ -434,9 +589,133 @@ fn make_merkle_branch(
+    ReedSolomon::new(num_data_shreds, num_coding_shreds)?.reconstruct(&mut shards)?;
+    let mask: Vec<_> = shreds.iter().map(Option::is_some).collect();
+    // Reconstruct code and data shreds from erasure encoded shards.
+    let mut shreds: Vec<_> = shreds
+        .into_iter()
+        .zip(shards)
+        .enumerate()
+        .map(|(index, (shred, shard))| {
+            if let Some(shred) = shred {
+                return Ok(shred);
+            }
+            let shard = shard.ok_or(TooFewShards)?;
+            if index < num_data_shreds {
+                let shred = ShredData::from_recovered_shard(&common_header.signature, shard)?;
+                let ShredCommonHeader {
+                    signature: _,
+                    shred_variant,
+                    slot,
+                    index: _,
+                    version,
+                    fec_set_index,
+                } = shred.common_header;
+                if shred_variant != ShredVariant::MerkleData(proof_size)
+                    || common_header.slot != slot
+                    || common_header.version != version
+                    || common_header.fec_set_index != fec_set_index
+                {
+                    return Err(Error::InvalidRecoveredShred);
+                }
+                Ok(Shred::ShredData(shred))
+            } else {
+                let offset = index - num_data_shreds;
+                let coding_header = CodingShredHeader {
+                    position: offset as u16,
+                    ..coding_header
+                };
+                let common_header = ShredCommonHeader {
+                    index: common_header.index + offset as u32,
+                    ..common_header
+                };
+                let shred = ShredCode::from_recovered_shard(common_header, coding_header, shard)?;
+                Ok(Shred::ShredCode(shred))
+            }
+        })
+        .collect::<Result<_, Error>>()?;
+    // Compute merkle tree and set the merkle branch on the recovered shreds.
+    let nodes: Vec<_> = shreds
+        .iter()
+        .map(Shred::merkle_tree_node)
+        .collect::<Result<_, _>>()?;
+    let tree = make_merkle_tree(nodes);
+    let merkle_root = &tree.last().unwrap().as_ref()[..SIZE_OF_MERKLE_ROOT];
+    let merkle_root = MerkleRoot::try_from(merkle_root).unwrap();
+    for (index, (shred, mask)) in shreds.iter_mut().zip(&mask).enumerate() {
+        if *mask {
+            if shred.merkle_root() != &merkle_root {
+                return Err(Error::InvalidMerkleProof);
+            }
+        } else {
+            let merkle_branch =
+                make_merkle_branch(index, num_shards, &tree).ok_or(Error::InvalidMerkleProof)?;
+            if merkle_branch.proof.len() != usize::from(proof_size) {
+                return Err(Error::InvalidMerkleProof);
+            }
+            shred.set_merkle_branch(merkle_branch)?;
+        }
+    }
+    // TODO: No need to verify merkle proof in sanitize here.
```

Comment:
still need verification?

Reply:
This needs to plumb some args to sanitize to skip merkle proof verification.
I will do that in a follow up PR.